src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -36,19 +38,107 @@ internal static bool IsEnabled()
-            new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
+        private async Task<HttpResponseMessage> SendInstrumentedAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Activity activity = null;
+            //check if user wants THIS request to be instrumented
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityName, request))
+            {
+                activity = new Activity(DiagnosticsHandlerLoggingStrings.ActivityName);
+
+                s_diagnosticListener.StartActivity(activity, new { Request = request });
+
+                //inject headers unless user does not want it to happen
+                //we may pass request to IsEnabled, but not clear how useful it is
+                if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityPropagateName))
+                {
+                    request.Headers.Add(HttpKnownHeaderNames.RequestId, activity.Id);
+                    
+                    //we expect baggage to be empty or contain a few items
+                    bool baggageIsEmpty = true;
+                    foreach (var _ in activity.Baggage)
+                    {
+                        baggageIsEmpty = false;
+                        break;
+                    }
+
+                    if (!baggageIsEmpty)
+                    {
+                        List<string> baggage = new List<string>();
+                        foreach (KeyValuePair<string, string> baggageItem in activity.Baggage)
+                        {
+                            baggage.Add(new NameValueHeaderValue(baggageItem.Key, baggageItem.Value).ToString());
+                        }
+                        request.Headers.Add(HttpKnownHeaderNames.CorrelationContext, baggage);
+                    }
+                }
+            }
+
+            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
+            try
+            {
+                await responseTask.ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                //we'll report task status in Activity.Stop
+                throw;
+            }
+            catch (Exception ex)
+            {
+                //If request was initialy instrumented, Activity.Current has all necessary context for logging
+                //If user decided to NOT instrument this request AND it threw an exception then:
+                //Activity.Current represents 'parent' Activity (presumably incoming request)
+                //So we let user log it as exception happened in this 'parent' activity
+                if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ExceptionEventName))
+                {
+                    s_diagnosticListener.Write(DiagnosticsHandlerLoggingStrings.ExceptionEventName, new {Exception = ex});
```

Comment:
@stephentoub 
As you noticed below, Activity and Request/Response events are mutually exclusive. Activity event consumers will always receive Exception with single exception object. Request/Response subscribers will always receive exception with guid and timestamp.

@cwe1ss 

1. From the logging perspective, URL may be logged in Request event along with Request-Id. And Exception will be logged with the same Id as well. But if subscriber did not instrument the request then in Exception callback there is no context for this particular request.
2. When exception event is received, Activity.Current holds context for this operation.
Subscriber can add any context to Activity.Current.Tags in the Request event and use them later in Exception and Response callbacks. It may cause Tags to overfill, considering that all tags should be written with every log message, it may be a problem.

Considering that Exception events should not happen too frequently, I'm fine with adding request to the payload.